lib/puppet/provider/netapp_net_port/cmode.rb

```diff
@@ -0,0 +1,51 @@
+require_relative '../netapp_cmode'
+
+Puppet::Type.type(:netapp_net_port).provide(:cmode, :parent => Puppet::Provider::NetappCmode) do
+  @doc = "Manage Netapp network port. [Family: cluster]"
+  
+  confine    :feature => :posix
+  defaultfor :feature => :posix
+
+  netapp_commands :netportlist => {:api => 'net-port-get-iter', :iter => true, :result_element => 'attributes-list'}
+  netapp_commands :netportmdfy => 'net-port-modify'
+
+  mk_resource_methods
+
+  def self.instances
+    Puppet.debug("Puppet::Provider::Netapp_net_port.cmode self.instances: Got to self.instances.")
+    results = netportlist() || []
+    netportinfos = []
+
+    results.each do |result|
+      node_name1 = result.child_get_string('node')
+      port_name1 = result.child_get_string('port')
+      node_port_name = node_name1 + "@" + port_name1
+
+      net_port_info_hash = {
+        :node_port_name => node_port_name,
+        :flowcontrol_admin => result.child_get_string('administrative-flowcontrol'),
+        :ensure            => :present
+      }
+      netportinfos << new(net_port_info_hash)
+    end
+    netportinfos
+  end
+
+  def self.prefetch(resources)
+    Puppet.debug("Puppet::Provider::Netapp_net_port.cmode: Got to self.prefetch.")
+    instances.each do |prov|
+      if resource = resources[prov.node_port_name]
```

Comment:
In this case, if the user supplies a title other than the node_port_name then the `self.provider` method would not be able to match providers to catalog resources. Eg:
```puppet
  netapp_net_port { 'alternate title':
    node_port_name    => 'vfiler-01@e0e',
    flowcontrol_admin => 'none',
  }
```
So the `self.prefetch` should match the `resource[:node_port_name]` to the `provider.name` as such:
```ruby
  def self.prefetch(resources)
    Puppet.debug("Puppet::Provider::Netapp_net_port.cmode: Got to self.prefetch.")
    net_ports = instances
    resources.each do |name, resource|
      if provider = net_ports.find { |net_port| net_port.name == resource[:node_port_name] }
        resources[name].provider = provider
      end
    end
  end
```

Reply:
site.pp
------- 

netapp_net_port{'abc@xyz':
  node_port_name    => 'hnidhi-vsim1@e0a',
  flowcontrol_admin => 'none'
}


We have confirmed that in:
------------------------------

def self.prefetch(resources)
  Puppet.debug("Puppet::Provider::Netapp_net_port.cmode: Got to self.prefetch.")
  net_ports = instances
  resources.each do |name, resource|
    p name
    if provider = net_ports.find { |net_port| net_port.name == resource[:node_port_name] }
       resources[name].provider = provider
    end
  end
end 


We printed name here and its seen that ‘name’ is having value as ‘node_port_name’s value only.
It’s not the alternate_title.
 
Hence,  
name printed as – hnidhi-vsim1@e0a

node_port_name is hnidhi-vsim1@e0a.

Hence its being matched with instances perfectly well, 
Kindly correct me if I am wrong in understanding this.


Rest code is changed as per suggestion.

Reply:
You are correct;`self.instances` changed `:node_port_name` to `:name` in https://github.com/puppetlabs/puppetlabs-netapp/pull/91/commits/4ee0d7e755087398292eb9df1144f225b63030e1 to fix the `No resource and no name in property hash` error. So we can make a more minor update here by just calling `prov.name` instead of `prov.node_port_name`:
```ruby
  def self.prefetch(resources)
    Puppet.debug("Puppet::Provider::Netapp_net_port.cmode: Got to self.prefetch.")
    instances.each do |prov|
      if resource = resources[prov.name]
        resource.provider = prov
      end
    end
```